### ISSUE TITLE
Links, relationships, discoverable actions, mapping to RFC5988

### ISSUE THREAD

**Issue by hhware** [json-api/json-api#622]
This proposal is a successor to #590. I would like to request to consider this for 1.0 because it won't make much sense afterwards.

@ethanresnick has submitted #618 while I was thinking about this. Even though the content of #618 is a part of what is proposed here, the scope of this issue is different from #618, so I am not hijacking #618...
##### Motivation
1. Five HTTP verbs are not enough to describe all possible actions available for a resource in a complex API. The solution is to add custom endpoints. There should be a way to associate those custom endpoints with the primary endpoint of the resource.
2. It would be great if JSON API allowed to build HATEOAS-compatible APIs. This requires ability to dynamically determine which actions can be taken on a given resource and its relationships, as well as what are the starting points to explore the API without much prior knowledge about it. It is often important for the clients to save trips to the server, so such information should be returned as part of a response, not on a separate endpoint.
3. Bring consistency to top-level and resource-level `links` and to relationships. A lot has been said on this. One thing I am trying to solve here is that each relationship provides multiple links, whereas members of top-level `links` and resource-level `self` are singular links, so they are not similar at all, one is a actually a component of another. Another aspect of it is making all `links` have the same structure, which can also be considered as a step towards the concept of "`links` can appear anywhere" (https://github.com/json-api/json-api/pull/588#issuecomment-98261811).
4. It would be useful to allow implementations to provide various links relevant to the given resource or response object, whether "verbs" or "nouns", whether representing actions on the endpoints of resource/its relationships/custom endpoints, or different aspects/views/versions/etc of the given resource, or anything else related to the given resource. Currently, only a very limited set of links is allowed (relationships, pagination, self), and `meta` is the only option for those who need more. It is workable, but it will drive fragmentation in the emerging JSON API ecosystem.
5. Provide an implementation of [linking in JSON](https://www.mnot.net/blog/2011/11/25/linking_in_json): allow association of [RFC5988 Link Relations](https://www.iana.org/assignments/link-relations/link-relations.xml) to _all_ links. This must be optional in order not to affect adoption of the spec.
6. Make links extendable: they are currently defined as strings, which does not allow to add any customization/context to them, either in basic spec in the future or via extensions today.
7. While achieving all this, the subset of the spec which defines minimally required data structures should not become much more complicated than it currently is, so that adoption is not affected.
##### Proposed changes
1. ~~Move relationships under resource-level key `relationships`, keeping `self` under resource-level key `links`. Relationships consist of _relationship objects_ (formerly link objects) keyed with relationship name (#618).~~ _[This is implemented in #618, #625]_
2. ~~Relationship objects consist of 3 members: `links`, `linkage`, `meta` (at least one is required).~~ _[This is implemented in #618, #625]_
3. ~~Member `links` has identical definition whether it occurs on top, resource, or relationship level.~~ _[This is implemented in #618, #625]_
4. Two forms of member `links` are possible: full (always available) or simplified (only available in special cases).
5. Full form of member `links`: an array, each element of it is a link object.
6. Define _link object_ in the spec as an object having the following members: 
   1. `name`: required, self-explanatory,
   2. `href`: required, self-explanatory, 
   3. `class`: required, classification of link type. It is a CSV-string with elements from the following set: `general,pagination,action,version,hierarchy,custom` (this set is obviously meant to be extendable in the future). Each link may have multiple classes. Elements of the set mean the following:
      1. `general` (could not find a good name for it): for generally applicable links (possible values: `self,related,type,copyright,help`),
      2. `pagination`: includes pagination members in top-level and relationship-level `links` (possible values: `prev,next,first,last`),
      3. `action`: actions on the given resource or relationships, available to the given user at the given moment of time, either at the resource's own endpoint or at some custom endpoint (possible values: `create,update,delete,explore`; general `self` substitutes the action of reading the current resource/relationship -- we might want to try to eliminate this inconsistency),
      4. `version`: links pointing at other versions of the given resource or providing versioning information about it (possible values: `version-history,working-copy,successor-version,predecessor-version,latest-version,latest-published-version`, see RFC5988),
      5. `hierarchy`: links for handling hierarchical content (possible values: `up`, see RFC5988),
      6. `custom`: all other (yet unclassified) links.
   4. `method`: optional, contains HTTP verb,
   5. `rfc5988`: optional, allows to add RFC5988 Link Relations. If defined, it should contain an object consisting of members keyed by names of `link-param`s from RFC5988 or its extensions. `URI-Reference` is considered to be set in `href`.
7. Attributes share namespace with relationships but not with links.
8. In each given member `links`, each link object is identified by a pair `name` and `class` (canonicalized, e.g. with elements sorted alphabetically), so such pairs must be unique. (Alternatively, `class` can be defined as array, e.g., `"class":["pagination","custom"]`: alphabetic reordering and serialization are needed for use as key.)
9. The spec should be able to reserve names of links of standard classes in order to have freedom for future extensions, but at the same time it would be useful if the spec allowed implementations to extend standard classes or define completely arbitrary links. To achieve that, the spec reserves possible ranges of values of `name` for link objects classified with one of `general,pagination,action,version,hierarchy`. Link with `class` ~~having any other value or any combination of values~~ _including any other value_ can have unrestricted `name`. E.g., if a link has `"class":"pagination"`, it could only have `name` as one of `prev,next,first,last`, but if it has `"class":"pagination,custom"` or just `"class":"custom"`, its `name` can be anything. All reserved values of `name` defined by spec must be unique, regardless of class. Proposed values per class to be defined by the spec are listed above in parentheses.
10. If member `links` only has link objects classified with single class, none of which is `custom`, and none of link objects has optional members, then such `links` can be returned by the server in simplified form: an object where values of `href` are keyed by values of `name` (`class` can be derived from `name` because `name` is unique for non-custom link classes). _As of #625,_ the simplified form is just ~~a tiny bit more complex than~~ the current version of the spec. Perhaps the full form can be described in a separate section of the base spec, not in the primary text flow of the spec. On another hand, support for full form could be an official extension, but IMHO it should be a part of the base spec.
11. In principle, link objects defined this way can be used anywhere in the response. E.g., the implementation can consider returning all attributes which contain URLs as link objects in full form.
12. Top-level link with `"class":"action"` and name `explore` could be used to communicate starting point for browsing a HATEOAS-compliant API.
##### Example

Illustrates all formats, both simple cases (article with ID=3) and complex cases (article with ID=2), dynamic permissions through actions (published article with ID=1 vs unpublished article with ID=2), RFC5988 mapping, custom actions (report for article with ID=1). 

I am not suggesting to allow mix of different link forms in one response, it is here only to shorten the example.

``` http
GET /articles?page=1&page_size=3&related_page_size=2

{
  "data": [{
    "type": "articles",
    "id": "1",
    "attributes": {
      "title": "JSON-LD paints my bikeshed!",
      "published": "true"
    },
    "relationships": {
      "authors": {
        "links": {
          "self": "http://example.com/articles/1/links/authors",
          "related": "http://example.com/articles/1/authors",
        },
        "linkage": [
          { "type": "authors", "id": "3" },
          { "type": "authors", "id": "4" }
        ]
      }
    },
    "links": [{
      "name": "self",
      "class": "general",
      "href": "http://example.com/articles/1"
    },{
      "name": "citation-report",
      "class": "custom",
      "method": "get",
      "href": "http://example.com/reports?type=citations&article_id=1",
      "rfc5988": {
        "rel": "http://example.com/notions/citation-reports",
        "type": "application/pdf",
        "title": "A report on citations of the given article over the last 5 years. 
Data is derived from ISI Web of Science database in Computer Science."
      }
    }]
  },{
    "type": "articles",
    "id": "2",
    "attributes": {
      "title": "JSON API paints my bikeshed!",
      "published": "false"
    },
    "relationships": {
      "authors": {
        "links": [{
          "name": "self",
          "class": "general",
          "href": "http://example.com/articles/2/links/authors",
          "rfc5988": {
            "rel": "self"
          }
        },{
          "name": "related",
          "class": "general",
          "href": "http://example.com/articles/2/authors",
          "method": "get",
          "rfc5988": {
            "rel": "related author"
          }
        },{
          "name": "update",
          "class": "action",
          "href": "http://example.com/articles/2/links/authors",
          "method": "patch"
        },{
          "name": "next",
          "class": "pagination",
          "href": "http://example.com/articles/2/authors?page=1&related_page_size=2",
          "method": "get",
          "rfc5988": {
            "rel": "related author next"
          }
        },{
          "name": "last",
          "class": "pagination",
          "href": "http://example.com/articles/2/authors?page=last&related_page_size=2",
          "method": "get",
          "rfc5988": {
            "rel": "related author last"
          }
        }],
        "linkage": [
          { "type": "authors", "id": "5" },
          { "type": "authors", "id": "12" }
        ],
        "meta": {
          "pagination": {
            "related_page_size": "2",
            "related_page_count": "3",
            "related_item_count": "5"
          }
        }
      }
    },
    "links": [{
      "name": "self",
      "class": "general",
      "href": "http://example.com/articles/2"
    },{
      "name": "update",
      "class": "action",
      "method": "patch",
      "href": "http://example.com/articles/2"
    },{
      "name": "delete",
      "class": "action",
      "method": "delete",
      "href": "http://example.com/articles/2"
    },{
      "name": "publish",
      "class": "action,custom",
      "method": "post",
      "href": "http://example.com/articles/2/publish"
    },{
      "name": "version-history",
      "class": "version",
      "method": "get",
      "href": "http://example.com/articles/2/version-history",
      "rfc5988": {
        "rel": "version-history"
      }
    },{
      "name": "version-drop-wip",
      "class": "version,action,custom",
      "method": "post",
      "href": "http://example.com/articles/2/version-drop-wip",
      "rfc5988": {
        "rel": "http://example.com/notions/version-drop-wip"
      }
    },{
      "name": "type",
      "class": "general",
      "href": "http://example.com/notions/articles",
      "rfc5988": {
        "rel": "type"
      }
    }]
  },{
    "type": "articles",
    "id": "3",
    "attributes": {
      "title": "On Silver Bullets",
      "published": "true"
    },
    "relationships": {
      "authors": {
        "links": {
          "self": "http://example.com/articles/3/links/authors",
          "related": "http://example.com/articles/3/authors"
        },
        "linkage": { "type": "authors", "id": "33" }
      }
    },
    "links": {
      "self": "http://example.com/articles/3"
    }
  }],
  "links": [{
    "name": "self",
    "class": "general",
    "href": "http://example.com/articles?page=1&page_size=3&related_page_size=2"
  },{
    "name": "explore",
    "class": "action",
    "href": "http://example.com/start-api-exploration"
  },{
    "name": "copyright",
    "class": "general",
    "href": "http://example.com/copyright",
    "rfc5988": {
      "rel": "copyright",
      "type": "text/html"
    }
  },{
    "name": "next",
    "class": "pagination",
    "href": "http://example.com/articles?page=2&page_size=3&related_page_size=2",
    "rfc5988": {
      "rel": "next"
    }
  },{
    "name": "last",
    "class": "pagination",
    "href": "http://example.com/articles?page=last&page_size=3&related_page_size=2",
    "rfc5988": {
      "rel": "last"
    }
  }],
  "meta": {
    "pagination": {
      "page_size": "3",
      "page_count": "15",
      "item_count": "43"
  }
}
```

**Comment by ethanresnick** [comment]
@hhware just wanted to say that I saw this and have been thinking about it. my big takeaway is that the core aim of your proposal—extending the `links` object to make it more powerful—is really important and spot on, but that your solution might be overly complex. i'll elaborate more on this when i can, but i want to write up my thoughts on #614 first, as that could effect this

**Comment by tkellen** [comment]
I really want JSON-API to point in the direction of HATEOAS too, but I think this proposal is too complex for 1.0, and honestly, for the relative near future of JSON-API.

That said, we're committed to merging #618 today or tomorrow, which gets us quite a bit closer.

Would you be willing to close this in favor of a PR based on the structure of #618 @hhware? I think the focus should be ensuring this approach is possible in an additive fashion.

**Comment by hhware** [comment]
@ethanresnick, @tkellen, thanks for your interest! At the time it was written, it was important to consider at least the items up to relationship-level `links` for 1.0, but since that is now handled as part of #625, this proposal becomes fully additive and naturally gets lower priority compared to other things. I have just "rebased" this onto #625 by simply removing the first items from the Proposed Changes and updating wording in one place. The example remains unchanged.

As for complexity -- I agree, but it covers multiple use cases and (hopefully) provides extensibility of links in the future. Also, I would like to emphasize that, since one of the main motifs of JSON API was to stay simple, one of the design goals for this proposal was to formulate changes in a way that two options are available. Those APIs which do not need advanced functionality can use the simplified option, which _is the base spec as of #625_.

As for the PR, of course I will be happy to provide it, but it would be great to hear _at least some constructive feedback_ on the proposed approach first. For instance, is the idea of classes acceptable at all? Could this go to the base spec or an extension is my ceiling?

I have changed bulleted list to enumerated list to simplify referencing.

> honestly, for the relative near future of JSON-API.

@tkellen, that is what scares me. "Full HATEOAS" may be a relatively exotic need, but discoverable actions (the main driving force behind this) is not a need of the future, it is a need of today: people want to build agile user interfaces which take into account current resource state and user permissions without requiring additional trips to the server. JSON API is a modern spec, coming of age this week, so IMHO, it should be addressing such challenges. Not that I am saying that my proposal is good, but I sincerely hope that some solution to the problem at hand can come out of it or at least with some help from it _in the very near future_.

**Comment by ethanresnick** [comment]
> As for the PR, of course I will be happy to provide it, but it would be great to hear at least some constructive feedback on the proposed approach first. For instance, is the idea of classes acceptable at all? Could this go to the base spec or an extension is my ceiling?

Ok, here goes some rough and only half-backed (hopefully) constructive criticism for possible simplifications:
- My inclination is that having classes of links add more complexity than it's worth. Categorizing link relations is, as your proposal mentions with `self`, difficult to do, and I'm not sure knowing a link's class is that helpful to a client that doesn't understand the specific link type (i.e. `name`). Also, letting one link be of multiple classes (which does seem like it would be necessary) then adds further conceptual overhead. Right now, the primary value of class seems to be that it carves out a space for user-defined link types (with the `"custom"` key), but I wonder if there are easier ways to do that.
- It seems weird to me to special case RFC5988 in it's own container. What I'd much rather do is merge `"name"` and the RFC5988 stuff into one value, which simply would be an (extended) relation.

Here's a very, very raw sketch that shows sort of the direction I'm thinking. (This example would be in a resource-level links, but the structure would apply to `links` anywhere.)

```
"links": {
  "self": "http://example.com/",
  "collection": ["http://example.com/type", "http://example.com/sub-type"],
  "http://example.com/rels/publish": "http://example.com/articles/2/publish"
}
```

This is a really simple, but complete, RFC 5988 implementation. The context for the link is whatever object the `"links"` is contained in; the relation is the key name (a string for registered relations and a url for extended ones); and the target document is the value. The nice thing about this is that all the keys used in the base spec already are appropriate, registered link relations, so we wouldn't have to change a thing (except indicate that, when they appear in a json-api document, the more-specific json-api meanings of each key also apply). And because the values can be arrays, we can support multiple links with the same relation type.

Moreover, the above could easily be extended to support curie-like aliasing to make things shorter, like:

```
"rel-prefixes": {
  "ex": "http://example.com/rels/"
},
"links": {
  "self": "http://example.com/",
  "collection": ["http://example.com/type", "http://example.com/sub-type"],
  "ex:publish": "http://example.com/articles/2/publish"
}
```

(I've calld this a "prefix" above only because I think curies is another one of those words that scares people.)

Finally, if we decide that it's _really_ necessary to hint at the supported methods, returned content types, etc. for each relation (and I'm not sure that it is—at least not in each response, though maybe we could in a JSON-HOME like entry point), we could replace the target URIs in the example above with objects, like so: 

```
"links": {
  "self": "http://example.com/",
  "http://example.com/rels/publish": {
    "href": "http://example.com/articles/2/publish", 
    "method": "post"
  }
}
```

What do you think? Could a more solid proposal be built on top of this? Are there any major cases it's overlooking? Etc.

**Comment by dgeb** [comment]
This is very interesting work @hhware. I think a PR to the base spec is premature at this point and would strongly prefer to see these HATEOAS concepts explored extensively as an extension. Of course, it would also be preferable if it could be an additive extension (or extensions) to the base spec. I think we should focus our work in the next few days on ensuring that this level of extensibility is possible.

**Comment by tkellen** [comment]
I think I was unclear in my previous comment--I meant a PR to ensure this kind of functionality is possible via additive changes later (basically what @dgeb just said).

**Comment by hhware** [comment]
@dgeb, thanks for encouraging note!

@ethanresnick, thanks for review! 

> having classes of links add more complexity than it's worth

I think extensibility is worth a lot. The primary use case: the client wants to know which actions are available on a resource, whether standard or custom -- just filter by class including `action`. The client wants to know which pagination links are available on a resource/relationship/response, whether standard or custom -- just filter by class including `pagination`. A few years from now, the client wants to know which links satisfy criterium `A` -- just filter by class `A`.

Link objects cannot be referenced directly, they are just part of the response, so I do not see why it is a problem to allow multiple classes per link, contrary to `type`, where multiple types per resource would generate quite a stir (even though they could probably help in cases like #571).

> I'm not sure knowing a link's class is that helpful to a client that doesn't understand the specific link type

In my view, ability for the client to filter links by criteria is the main reason why classes are introduced. If the client does not need to filter, then it can of course just find the desired link by name.

> Right now, the primary value of class seems to be that it carves out a space for user-defined link types (with the "custom" key), 

I added a link to the example above to illustrate another case: `version-drop-wip`. Suppose some API allows to maintain a WIP version of a resource along with a published version, and hence needs actions to publish it (i.e., replace published version with WIP) and cancel editing (i.e., drop WIP). The spec does not make provisions for this scenario, so it has to be `custom`. Is it action? Yes, so it has to have `action`. Is it related to versioning? Yes, it has to have `version`.

> but I wonder if there are easier ways to do that.

If it were only for `custom`, then obviously it would be easier to add a boolean flag to indicate it. Having separate flags for actions, pagination, etc, does not seem very practical. Again, I see it as an extensible mechanism for classifying links, which could be used by the server and client for various purposes.

> It seems weird to me to special case RFC5988 in it's own container. 

I do not mind putting them directly into the link object, but I thought that it would be easier to store them in their own container because they are defined by different specs, so there is a potential for parameter name conflict. 

> What I'd much rather do is merge "name" and the RFC5988 stuff into one value

I do not think it is practical. This proposal provides the ability for the API to define links with arbitrary names. RFC5988 defines a limited set of readable names, and the rest have to be URLs. I doubt many people would actually want to use URLs for names because they are not very readable and might hinder communication. I think that my example with action `publish` is much more readable than your example with `"http://example.com/rels/publish": "http://example.com/articles/2/publish"`.

> This is a really simple, but complete, RFC 5988 implementation

I do not think it is complete, because it does not allow, for example, to pass `media` or `type` of RFC5988. 

> Finally, if we decide that it's really necessary to hint at the supported methods, returned content types, etc. for each relation

In my view, that would be closer to support of RFC5988.

> we could replace the target URIs in the example above with objects, like so:

This is what I am proposing, but not that much for RFC5988 compliance, mainly for (1) ability to add `class`, (2) extensibility. 

The main difference between our proposals at this point is how links are identified. You suggest to key them by name, but what if a particular API defines a custom link name, and JSON API decides to reserve it a few years later? Adding class for custom links allows to avoid such conflicts, while allowing people to assign convenient meaningful names to links (not ugly constructs like `http://example.com/rels/publish`). My proposal is closer to how `included` is structured: an array with a bunch of heterogeneous objects.

**Comment by ethanresnick** [comment]
> I doubt many people would actually want to use URLs for names because they are not very readable and might hinder communication

I agree that URIs as keys are ugly, but prefixes do ameliorate this to a large extent. And we can get even more clever than my example earlier. For example, we could define the `api` prefix as automatically mapping to the `{current host}/rels/`. So an api at `http://example.com/v2` could return a resource-level `links` that looks like this:

```
"links": {
  "api:publish": "http://example.com/v2/articles/2/publish"
}
```

And `api:publish` would be mapped to the url "http://example.com/rels/publish" without the user having to define `http://example.com/rels/` as the value of a prefix anywhere. (And this works even if the user doesn't ultimately bother to put anything at the `http://example.com/rels/publish` URI.)

The above still isn't as clean as a simple `"publish"` key, it's true. But, if we have a `name`, we still have to put the registered relation name or (something that maps to) a relation URI _somewhere_ for us to have RFC 5988 compliance. And, to me, whatever ugliness is left in the `api:publish` key is preferable to having both a `name` and a relation mapping, which seems redundant. I don't like the idea of links with keys like `"name": "first", "rel": "first"`, `"name": "self", "rel": "self"`, etc.
&nbsp;

> You suggest to key them by name, but what if a particular API defines a custom link name, and JSON API decides to reserve it a few years later? 

Future JSON-API keys might be extended relations, in which case they'd live at `http://jsonapi.org/rels/whatever` and couldn't conflict with user-defined keys (which live on other domains). The only possibility for conflict, therefore, would be if JSON API later defines a more-specific meaning for an existing registered link relation, and the user was already using that relation in their API with the generic meaning (but not the json-api–specific specialized meaning). But that possibility could be prevented by the right reasoning in #614.
&nbsp;

> The main difference between our proposals at this point is how links are identified... My proposal is closer to how `included` is structured: an array with a bunch of heterogeneous objects.

Yup! And you're right that your structure is closer to `included`. But it's also less similar to the current/"simplified" form of a `links` object that's in the base spec. I wanted the more complex link format to feel like it "grows out of" the base spec's simplified form, which is why I structured my proposal as I did. I think that's preferable to `links` sometimes being an object and sometimes being an array, which feels more like two competing formats, rather than one unified one that is extended as necessary. Under my proposal, `links` is always an object; the relation is always the key; and the value is made up of a set of "building blocks", namely:
1. A "link target" object, which must contain an `"href"` key and can contain other keys, like `method`, `content-type`, etc. An optional `class` key could go here too; I'm still not 100% sold on it (namely, because I worry we'd get the taxonomy wrong), but it would fit into the structure perfectly.
2. A "link target" object can be expressed more compactly as just a string, which equates to an object with just an href member.
3. And the value of the link is a single "link target" object/string, or an array of such objects.

**Comment by hhware** [comment]
> is preferable to having both a name and a relation mapping, which seems redundant. 

It is redundant only if they always coincide. What if an API had a naming convention for actions and did not want to follow suggestions of IANA for whatever reason? What if an API did not want to use URLs for custom link names? Even if it were not the case and the valuess were the same, some redundancy can be tolerated because RFC5988 is external to JSON API.

> Future JSON-API keys might be extended relations, in which case they'd live at http://jsonapi.org/rels/whatever and couldn't conflict with user-defined keys (which live on other domains)

I personally really dislike the idea of an URL as a link name. Also, if two different domains are used, your argument "without the user having to define `http://example.com/rels/` as the value of a prefix anywhere" does not seem to apply.

> And the value of the link is a single "link target" object/string, or an array of such objects.

I did not think of an array of links as a value. Why would the client need multiple locations for one link name, are there example use cases? To me, it seems like duplication of functionality of to-many relationships at link object level. IMHO, the semantics of the link object is a location where something (defined by the link name) can be found or done, and if there are multiple locations, they should have distinct link names.

> I agree that URIs as keys are ugly, but prefixes do ameliorate this to a large extent.

OK, so there is a choice between (a) `included`-like array and (b) current-spec-like object with namespace for custom links. I personally still prefer the option (a) because it allows implementations to select arbitrary names for links. The spec so far was able to avoid going in the way of implementations by not imposing naming conventions, I feel that link is an important concept which deserves the same treatment. However, I understand the appeal of the option (b), and if reasonable namespacing convention is defined ("`http://example.com/rels/` as the value of a prefix" does not seem reasonable to me), it could be acceptable too. The convention of [dasherized custom link names](https://github.com/json-api/json-api/issues/602#issuecomment-103310211) will not work because IANA registry already contains dasherized names, so it seems it has to be a prefix. 

If the option (b) is adopted, there is no need to append `custom` to the value of `class` _if_ the link is meant to supplement an existing class. (If option (a) is adopted, appending of `custom` to existing class can be avoided by adding a separate boolean flag member to link objects to indicate whether it is a custom link or not.)

> HATEOAS concepts explored extensively as an extension. Of course, it would also be preferable if it could be an additive extension

@dgeb, the only thing which concerns me with formulating this particular proposal as extension is that it would be substituting the definition of `links` as an object in the base spec with an array in the extension (or, in the framework of @ethanresnick's suggestion, the definition of value as string in the base spec is substituted with an object in the extension). I do not think this problem exists the other way around: as proposed here, the base spec defines 2 forms, full and simplified, so the implementation should be handling both. But if the base spec defines a member as something and an extension redefines it as something else, it cannot actually be classified as additive (it will fall under (1) in @ethanresnick's [classification](https://github.com/json-api/json-api/issues/614#issuecomment-103209872)). Rich links are not possible as additive extension simply because the entity to be extended is a string, and strings cannot be extended. The only additive way which occurs to me is something ugly like 

```
"links": {
  "self": "http://example.org/arcicles/1",
  "next": "http://example.org/arcicles/1/next",
  "supplementary": {
    "self": {
      "key1": "v11",
      "key2": "v12"
    },
    "next": {
      "key1": "v21",
      "key2": "v22"
    }
  }
}
```

IMHO, this stuff belongs to the base spec, possibly to a separate section or something, but not to an extension.

**Comment by tkellen** [comment]
> @tkellen, that is what scares me. "Full HATEOAS" may be a relatively exotic need, but discoverable actions (the main driving force behind this) is not a need of the future, it is a need of today: people want to build agile user interfaces which take into account current resource state and user permissions without requiring additional trips to the server. JSON API is a modern spec, coming of age this week, so IMHO, it should be addressing such challenges. Not that I am saying that my proposal is good, but I sincerely hope that some solution to the problem at hand can come out of it or at least with some help from it in the very near future.

I agree that this is **extremely** important stuff. That being said, I also think we need to give the developer community time to absorb and implement the concepts we're introducing before we layer on yet more.

What we are proposing with JSON-API is already a far cry from the "expose your database tables over the wire" mentality that pervades API design today. I strongly believe that we need to be tactical in our approach if we are to succeed in garnering wide adoption for these ideas.

As long as we leave the door open for this sort of functionality to be introduced additively, I feel we're in great shape to create a new baseline for API design on the web.

I think I speak for all of my co-authors when I say that we're committed HATOES, just not at the expense of alienating the wider developer community.

**Comment by tkellen** [comment]
In short, please bear with us. We're headed in the same direction, just at different speeds. I think the best thing you can do **right now** is focus your efforts on making sure we don't introduce roadblocks that will prevent our progression in the future.

**Comment by dgeb** [comment]
Well said, @tkellen - I agree completely.

**Comment by steveklabnik** [comment]
> I think I speak for all of my co-authors when I say that we're committed HATOES, just not at the expense of alienating the wider developer community.

Yes, absolutely.

**Comment by tkellen** [comment]
> Rich links are not possible as additive extension simply because the
> entity to be extended is a string, and strings cannot be extended.

I think this could be resolved with a focused PR that adds support for an
expanded form of links as an object (which would typically not be used
initially, but would normatively prepare implementations to handle this).

**Comment by hhware** [comment]
> you can do **right now** is focus your efforts on making sure we don't introduce roadblocks that will prevent our progression in the future

Sure, this is the primary reason why this PR has been submitted before 1.0...

> I think this could be resolved with a focused PR that adds support for an
> expanded form of links as an object (which would typically not be used
> initially, but would normatively prepare implementations to handle this).

Agree, but that means that the (a) vs (b) from https://github.com/json-api/json-api/issues/622#issuecomment-103685484 has to be decided before 1.0. Opinions?

**Comment by tkellen** [comment]
I would say B. I think we should start by specifying the bare minimum "expanded" link form to support current functionality. The base spec should normatively make these two forms interchangeable, and perhaps nothing more needs to be said right now.

``` json
"links": {
  "self": "http://example.com/posts",
}
```

``` json
"links": {
  "self": {
    "href": "http://example.com/posts"
  }
}
```

It's late and I'm probably missing something, but that's the sort of thing I'm imagining.

**Comment by tkellen** [comment]
We could perhaps reserve the entire namespace inside the object to start, save for `meta`, for people who want to explore this sort of thing.

**Comment by dgeb** [comment]
I think allowing objects for links, reserving the whole namespace, and only recognizing `meta` and `href` is the way to go. Big :+1:

**Comment by ethanresnick** [comment]
@tkellen Yup, I agree completely! Like I said earlier, I proposed (b) over (a) only because it seemed to grow more naturally out of what we have now. That isn't meant as a knock on @hhware's proposal—I think it's asking all the right questions and the primary difference we're talking about is stylistic anyway—I just think approach (b) feels like a more minimal change.

I'd only add that, to preserve the option of later mapping the keys to RFC 5988 relations, we should account in the base spec for the fact that it's quite common for multiple links to have the same relation name (e.g. the `profile` and `collection` relations, both of which are directly applicable to JSON API).

I think we could account for that either by allowing or mandating that the value at each key can be an array. Do either you, @hhware, or @dgeb have a preference one way or the other? Mandating an array means one less format for clients to parse, but it's a bit uglier if the array only has one item.

That is, which one:

```
{
  "links": {
    "self": "http://example.com/people/1",
    "profile": ["http://example.com/my-awesome-ext", "http://jsonapi.org/awesome-standard-ext"]
  }
}
```

Or

```
{
  "links": {
    "self": ["http://example.com/people/1"],
    "profile": ["http://example.com/my-awesome-ext", "http://jsonapi.org/awesome-standard-ext"]
  }
}
```

**Comment by dgeb** [comment]
I feel that allowing the array form is premature at this point and that the object form could encompass such a structure if needed.

**Comment by tkellen** [comment]
Agreed.

**Comment by ethanresnick** [comment]
@dgeb @tkellen Ok, that's fair. (Though #614 might change the needs here.) Out of curiosity, though, how would the object form accommodate multiple links? The solutions I can think of quickly are pretty ugly.

**Comment by dgeb** [comment]
> Out of curiosity, though, how would the object form accommodate multiple links?

I really think it's premature to answer this now but: a member of the object (not sure of the key) could contain the array - or perhaps another object - that represents multiple links.

**Comment by ethanresnick** [comment]
@dgeb Ok, yeah, that's more or less what I was thinking, and that's fine (again, pending #614). 

Could you, @tkellen, and @steveklabnik actually jump back into #614 when you have a sec? I've been publicly brainstorming some stuff there, and am working it into a proposal, but it'd be helpful to know if you guys think I'm on the right track or think there's anything missing from my conceptual structure before I try to work up something more specific.

**Comment by tkellen** [comment]
@hhware Would you be willing to put together a PR to cover this bare minimum use-case? Do you feel like this is enough to pave the way forward for real support of this sort of thing eventually?

**Comment by hhware** [comment]
> Do either you, @hhware, or @dgeb have a preference one way or the other?

Sorry, @ethanresnick, I really do not understand the point of multiple links under the same name, so my preference is not to support arrays :) IMHO, the notion of classes allows to achieve the same result in a more versatile way. In your example, the API returns 2 profiles. Why not have a link class `profiles`, assign it to both links and name them with distinct meaningful names? Both of them could have RFC5988 relation name `profile`, but if JSON API link name is different from RFC5988 relation name, the client can still distinguish each link by its JSON API link name. How can each profile be referred to in your example? 

Also, suppose 5 profiles in total are supported by the server, and the given API response returns 2 of them. How can the client know which of 2 URLs matches which profile if the URLs do not have names? Is the client supposed to parse the profile URL and figure out which URL corresponds to which profile? I feel like the only use for arrays is "load balancing", but this should be a responsibility of a layer lower in the stack...

> @hhware Would you be willing to put together a PR to cover this bare minimum use-case? 

Sure, I will prepare a PR tomorrow.

> Do you feel like this is enough to pave the way forward for real support of this sort of thing eventually?

Yes, I think so. As I mentioned above, the only limitation I see now is the imminent requirement for a namespace for custom links, which is not in line with the approach taken by JSON API in other areas. Nevertheless, I believe it is perfectly workable.

**Comment by ethanresnick** [comment]
@hhware I think we're talking past each other a bit. The key difference seems to be that your proposal has a required `name` property for each link with an optional RFC 5988 relation, whereas my proposal has a required RFC 5988 relation without (necessarily) giving each link a `name`.

<a id="named-collections"></a> Now, let me say right off the bat that I do think there are some use cases for giving a link a `name`, separate from its `href` or relation. For example, the `collection` relation lets a resource link to the collection(s) that it belongs to. If links to these containing collection(s) are going to be shown directly to users, it might make sense to give them human-readable names. In my proposal, that could happen by adding `name` as an extra, optional property in the link's object, like this:

```
"links": {
  "collection": {"href": "http://example.com/people", "name": "People Resources"}
}
```

But I think the example of profiles (below) shows clearly why it makes sense to make the relation the required part and the `name` the optional part, rather than vice-versa.

> In your example, the API returns 2 profiles. Why not have a link class `profiles`, assign it to both links and name them with distinct meaningful names?... How can the client know which of 2 URLs matches which profile if the URLs do not have names? Is the client supposed to parse the profile URL and figure out which URL corresponds to which profile?

In the profile specification, a profile's URI _is_ its name. The whole idea is that, by using a URI as a name, it can be globally identifiable/unique. So, yes, the client is supposed to match on the URI—that's how a generic client knows whether it actually understands the specified profile—and, again, having another name would just be redundant.

Imagine a client that comes across an API and gets back this JSON:

```
"links": [{
    "name": "semi-standard-immutables-documentation",
    "class": "profiles",
    "href": "http://popular-jsonapi-extension.org/immutables",
    "rfc5988": {
      "rel": "profile"
    }
  }, {
    "name": "semi-standard-write-template",
    "class": "profiles",
    "href": "http://popular-jsonapi-extension.org/write-template",
    "rfc5988": {
      "rel": "profile"
    }
  }, 
  // more links here...
]
```

There are two profiles here. They're semi-standard in that they're published publicly on some domain, such that a generic client might actually know how to interpret the additional semantics they convey. But they're not part of the base spec (in which case a profile isn't necessary) or specific to the `example.com` API. (Any semantics specific to a given API could be linked as profiles, but don't usually need to be, since a generic client won't know how to process the extra semantics anyway, and a client written specifically for the given API likely knows that the profile is in effect without it being declared.)

So, let's say a generic client comes across the JSON above. The client knows that it can understand the `http://popular-jsonapi-extension.org/write-template` profile, which it identifies internally by that URI but not by `name` (because the APIs it's interacting with could name it anything). So, it gets this blob of JSON and it identifies the profiles using the `profiles` class, which just ends up duplicating the `rel`, and then it throws away the `name` entirely and only looks at the `href` anyway. So this whole structure ends up communicating no more to a generic client than simply:

```
"links": {
    "profile": ["http://popular-jsonapi-extension.org/immutables","http://popular-jsonapi-extension.org/write-template"]
  }
```

And, again, a client written just for this API (as opposed to a generic client) probably already knows which profiles are in effect, so it doesn't even need to look at `links`. And, in the rare cases it does need to look at links, it would be better off identifying the profiles there by URI anyway. And, finally, if for some reason that's still inadequate, an optional name could always be added:

```
"links": {
    "profile": [{"href": "http://popular-jsonapi-extension.org/immutables", "name": "whatever", {"href": "http://popular-jsonapi-extension.org/write-template", "name": "profile 2"}]
}
```

But hopefully the above shows the key point: the problem with `name` is that it's not globally unique—that is, different APIs can use the same name for different things—whereas a link relation's name=>meaning mapping is unique. This is what allows generic clients to extract meaning from the link relations. So identifying the URIs by relation is key. And, once you do that, it's quite clear why arrays are needed, namely: because you can't just change the key to make each one distinct; you have to use the standard off-the-shelf relation name, and multiple links can share the same relation.

Meanwhile, even if the only use case for `links` in a given API is custom actions that don't have off-the-shelf link relations that describe them (and this won't always be the only, or even primary, use case), keying those actions with strings like `api:publish` still works fine; the fact that `api:publish` can be converted to a URI and therefore an extended link relation becomes an incidental detail that doesn't harm anyone (again, the API creator doesn't even have to put anything at the URI), but that allows the benefits of merging the relation name and `name` in cases where they'd otherwise be the same thing.

**Comment by hhware** [comment]
> So this whole structure ends up communicating no more to a generic client than simply

@ethanresnick, got it. Indeed, it seems to be a very reasonable shortcut when relation name is used for link object key.

Well, perhaps we should indeed allow arrays as values of `href`? Unless we do it now, it would be awkward, e.g., to have `href` for a single URL and `hrefs` for multiple URLs, and the only way to have both under `href` is to define the whole thing in extension, not in the base spec. Since the whole purpose of this is to leave the door wide open for the future, it should not be a problem to partly close it later if the array use case ends up not being supported. This is accounted for in my PR #644, so @dgeb and @tkellen, please let me know if it needs to be removed.

**Comment by tkellen** [comment]
Closing in favor of #644.  Thanks @hhware!
